osfbot/exts/moderation/logs.py

```diff
@@ -0,0 +1,243 @@
+from disnake.ui import Button
+from disnake.utils import format_dt
+from loguru import logger
+
+from ...bot import Bot
+from ...constants import Channels, Colors
+
+
+class ModerationLogs(Cog):
+    """Used to log infractions and important actions in the specified channel."""
+
+    def __init__(self, bot: Bot):
+        self.bot = bot
+        self.log_channel: Optional[TextChannel] = None
+        super().__init__()
+
+    async def post_message(
+        self,
+        embed: Embed,
+        components: List[Component] = None,
+    ) -> Optional[Message]:
+        """Send the given message in the log channel."""
+        if not self.log_channel:
+            await self.bot.wait_until_ready()
+            self.log_channel = await self.bot.fetch_channel(Channels.log)
+
+            if not self.log_channel:
+                logger.error(f"Failed to get log channel with ID ({Channels.log})")
+
+        return await self.log_channel.send(embed=embed, components=components)
+
+    async def post_formatted_message(
+        self,
+        actor: Member,
+        title: str,
+        color: int = Colors.green,
+        fields: Optional[list] = None,
+        url: Optional[str] = None,
+        components: List[Component] = None,
+    ) -> None:
+        """Formats an embed to be posted in the log channel."""
+        embed = Embed(
+            title=title,
+            color=color,
+            timestamp=datetime.now(),
+        ).set_thumbnail(url=actor.display_avatar.url)
+
+        if url:
+            embed.url = url
+
+        for field in fields:
+            embed.add_field(
+                name=field["name"], value=field["value"], inline=field["inline"]
+            )
+
+        await self.post_message(embed=embed, components=components)
+
+    @Cog.listener()
+    async def on_raw_message_delete(self, payload: RawMessageDeleteEvent) -> None:
+        """Logs message deletion."""
+        if message := payload.cached_message:
+            if message.author.bot:
+                return
+
+            fields = [
+                {
+                    "name": "Author",
+                    "value": f"{message.author.mention} **({message.author.name}) ({message.author.id})**",
+                    "inline": False,
+                },
+                {
+                    "name": "Channel",
+                    "value": f"{message.channel.mention} **({message.channel.name}) ({message.channel.id})**",
+                    "inline": False,
+                },
+                {
+                    "name": "Content",
+                    "value": f"```{message.content[:1000]+('...' if len(message.content) > 1024 else '')}```",
+                    "inline": False,
+                },
+            ]
+
+            await self.post_formatted_message(
+                actor=message.author,
+                title=f"Message Deleted ({message.id})",
+                color=Colors.red,
+                fields=fields,
+            )
+
+        else:
```

Comment:
I don't think it is useful to log this as the mod cannot see the content, unless there is a way to do so.

Reply:
I can add more information via the `payload`, but I believe this would be useful when a certain member is deleting multiple uncached messages.

Again, details about the member aren't currently logged. I'll try something out.

Reply:
Alright, try to add details about the author of the message.

Reply:
Turns out the `payload` object does not provide much information for an uncached message, other than the channel and message ID. I still think it's a good idea to keep this for a log - uncached messages are only those that are older than 2 weeks, or sent before the bot was up.